adafruit_midi/__init__.py

```diff
@@ -55,39 +58,100 @@ class MIDI:
+                 in_channel=None, out_channel=0, debug=False, in_buf_size=30):
         self._midi_in = midi_in
         self._midi_out = midi_out
-        self._in_channel = in_channel
-        self._out_channel = out_channel
+        self.in_channel = in_channel
+        self.out_channel = out_channel
         self._debug = debug
+        # This input buffer holds what has been read from midi_in
+        self._in_buf = bytearray(0)
+        self._in_buf_size = in_buf_size
         self._outbuf = bytearray(4)
+        self._skipped_bytes = 0
 
     @property
     def in_channel(self):
         """The incoming MIDI channel. Must be 0-15. Correlates to MIDI channels 1-16, e.g.
-        ``in_channel(3)`` will listen on MIDI channel 4. Default is 0."""
+        ``in_channel(3)`` will listen on MIDI channel 4.
+        Can also listen on multiple channels, e.g. ``in_channel((0,1,2))``
+        will listen on MIDI channels 1-3 or ``in_channel("ALL")`` for every channel.
+        Default is None."""
         return self._in_channel
 
+    # pylint: disable=attribute-defined-outside-init
     @in_channel.setter
     def in_channel(self, channel):
-        if channel is not None and not 0 <= channel <= 15:
+        if channel is None or (isinstance(channel, int) and 0 <= channel <= 15):
+            self._in_channel = channel
+        elif isinstance(channel, str) and channel == "ALL":
+            self._in_channel = ALL_CHANNELS
+        elif isinstance(channel, tuple) and all(0 <= c <= 15 for c in channel):
+            self._in_channel = channel
+        else:
             raise RuntimeError("Invalid input channel")
-        self._in_channel = channel
 
     @property
     def out_channel(self):
         """The outgoing MIDI channel. Must be 0-15. Correlates to MIDI channels 1-16, e.g.
         ``out_channel(3)`` will send to MIDI channel 4. Default is 0."""
         return self._out_channel
 
+    # pylint: disable=attribute-defined-outside-init
     @out_channel.setter
     def out_channel(self, channel):
         if not 0 <= channel <= 15:
             raise RuntimeError("Invalid output channel")
         self._out_channel = channel
 
+    def receive(self):
+        """Read messages from MIDI port, store them in internal read buffer, then parse that data
+        and return the first MIDI message (event).
+
+        Returns (MIDIMessage object, channel) or (None, None) for nothing.
+        """
+        ### could check _midi_in is an object OR correct object OR correct interface here?
+        # If the buffer here is not full then read as much as we can fit from
+        # the input port
+        if len(self._in_buf) < self._in_buf_size:
+            bytes_in = self._midi_in.read(self._in_buf_size - len(self._in_buf))
+            if bytes_in:
+                if self._debug:
+                    print("Receiving: ", [hex(i) for i in bytes_in])
+                self._in_buf.extend(bytes_in)
```

Comment:
I did start a discussion with @dhalbert about this in Discord and also pondered some sort of fixed size buffer and tracking what was in use to avoid reallocations.

BTW, is `midi_in` always going to have a `readinto`? There was some discusion about using this with UARTS.

Reply:
I actually did some optimizations on it last night. They are incomplete for now but here: https://github.com/adafruit/Adafruit_CircuitPython_MIDI/commit/e63f0170f9759beca0b9b24870c6e92ecff4e583

Let's not worry about it for this PR. We can do subsequent PRs to optimize.

Reply:
Here's the previous Discord chat:

> 23:20
]
kjw
:
hello all. Is anyone aware of any code in CircuitPython which manipulates a bytearray buffer? The CP implementation is very bare bones with append() and extend() only.
[
23:21
]
kjw
:
I'm wondering how to (efficiently) manage a bytearray where I need to remove a little bit from the beginning
[
23:27
]
kjw
:

>>> B = bytearray([4,5,6,7,8,9])
>>> del B[0]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'bytearray' object does not support item deletion

[
23:27
]
kjw
:
:frowning:
[
23:28
]
madbodger
:
Use slice notation?

B[1:]
bytearray(b'\x05\x06\x07\x08\t')

[
23:29
]
kjw
:
I'd like to efficiently modify an existing one
[
23:30
]
danh
:
The internal storage doesn't allow trimming off the front, I believe, so you'll need to copy one way or another
[
23:30
]
kjw
:

>>> B
bytearray(b'\x04\x05\x06\x07\x08\t')
>>> B = B[2:]
>>> B
bytearray(b'\x06\x07\x08\t')

[
23:30
]
danh
:
is array.array of bytes more functional for you?
[
23:31
]
danh
:
that's why various read and write functions take boundaries for buffers so you can do partial reads or writes
[
23:32
]
kjw
:
I'm buffering up some MIDI data. I was thinking of holding a certain amount of it in the MIDI class.
[
23:32
]
kjw
:
What do you think is best, I've not looked at array.array. Is that going to be bigger because it's int rather than bytes?
[
23:32
]
danh
:
could you just keep start and stop indices, or do a circular buffer?
[
23:33
]
danh
:
array.array is what type you say: it could be bytes, 16-bit words, or 32-bit-words
[
23:33
]
kjw
:
I was pondering that but thought I'd start simple!
[
23:34
]
danh
:
you could keep a list of bytes or byte arrays and join them at the end. b''.join(...) is the "efficient" way to concatenate
[
23:34
]
danh
:
(assuming that works, haven't tried it)
[
23:35
]
kjw
:
ok, i'll start with just over writing with slices and leave a note to revisit it if memory/performance is grim
[
23:35
]
kjw
:
thanks for the ideas